File: core/src/com/bosowski/oslark/screens/CharacterSelectionScreen.java
package com.bosowski.oslark.screens;

import com.badlogic.gdx.Game;
import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.GL20;
import com.badlogic.gdx.math.Rectangle;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.math.Vector3;
import com.badlogic.gdx.scenes.scene2d.InputEvent;
import com.badlogic.gdx.scenes.scene2d.ui.Label;
import com.badlogic.gdx.scenes.scene2d.ui.TextButton;
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener;
import com.bosowski.oslark.World;
import com.bosowski.oslark.gameObjects.Player;
import com.bosowski.oslark.gameObjects.Terrain;
import com.bosowski.oslark.generation.areas.Dungeon;
import com.bosowski.oslark.generation.areas.DungeonCell;
import com.bosowski.oslark.generation.areas.DungeonRoom;
import com.bosowski.oslark.generation.areas.Maze;
import com.bosowski.oslark.main.Session;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Vector;

public class CharacterSelectionScreen extends AbstractGameScreen {

    private final JSONArray characterArray;

    public CharacterSelectionScreen(Game game) {
        super(game);
        String userJson = null;
        try {
            userJson = Session.instance.loadUser();
        } catch (IOException e) {
            e.printStackTrace();
        }
        System.out.println(userJson);
        JSONObject obj = new JSONObject(userJson);
        System.out.println(obj.get("username"));
        characterArray = obj.getJSONArray("characters");
        setUpUI();
    }

    @Override
    protected void setUpUI(){
        Label label = new Label("Choose your character", fieldSkins);
        label.setPosition(Gdx.graphics.getWidth() / 3, Gdx.graphics.getHeight() - 100);
        stage.addActor(label);

        LinkedHashMap<String, Player> characters = new LinkedHashMap<>();

        int i = 0;
        for (Object character : characterArray) {
            Player player = new Player((JSONObject) character);
            TextButton characterButton = new TextButton(player.getName() + "(level " + player.getLevel() + " " + player.getCharacterClass(), fieldSkins);
            characters.put(player.getName(), player);
            characterButton.setPosition(label.getX(), label.getY() - 50 * (i + 1));
            characterButton.addListener(new ClickListener() {
                @Override
                public boolean touchDown(InputEvent event, float x, float y, int point, int button) {
                    System.out.println("Picked " + player.getName());
                    World.instance.setPlayer(characters.get(player.getName()));
                    String worldJson = null;
                    try {
                        worldJson = Session.instance.loadWorld("characterName=" + World.instance.getPlayer().getName());

                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                    System.out.println(worldJson);
                    JSONObject jsonObj = new JSONObject(worldJson);
                    JSONArray terrain = jsonObj.getJSONArray("terrain");
                    for (Object tile : terrain) {
                        Terrain terrainTile = new Terrain((JSONObject) tile);
                        World.instance.instantiate(terrainTile);
                        System.out.println("Loaded terrain: " + terrainTile);
                    }

                    game.setScreen(new GameScreen(game));
                    return true;
                }
            });
            stage.addActor(characterButton);
            i++;
        }
    }

    @Override
    public void render(float deltaTime) {
        // Clear the buffer
        Gdx.gl.glClearColor(0, 0, 0, 1);
        Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT);

        stage.act(deltaTime);
        stage.draw();
    }

    @Override
    public void resize(int width, int height) {
        super.resize(width, height);
    }

    @Override
    public void show() {
        Gdx.input.setInputProcessor(stage);
    }

    @Override
    public void hide() {

    }

    @Override
    public void pause() {

    }
}

File: OslarkGame/core/src/com/bosowski/oslark/screens/CharacterSelectionScreen.kt
package com.bosowski.oslark.screens

import com.badlogic.gdx.Game
import com.badlogic.gdx.Gdx
import com.badlogic.gdx.graphics.GL20
import com.badlogic.gdx.scenes.scene2d.InputEvent
import com.badlogic.gdx.scenes.scene2d.ui.*
import com.bosowski.oslark.managers.NetworkManager
import org.json.JSONArray
import org.json.JSONObject
import java.io.IOException
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener
import com.bosowski.oslark.Assets
import com.bosowski.oslark.World
import com.bosowski.oslark.enums.State
import com.bosowski.oslark.gameObjects.GameObject

class CharacterSelectionScreen(game: Game) : AbstractGameScreen(game) {
  private val TAG = CharacterSelectionScreen::javaClass.name

  private val characterArray: JSONArray

  init {
    World.game = game
    var userJson: String? = null
    try {
      userJson = NetworkManager.instance.loadUser()
    } catch (e: IOException) {
      e.printStackTrace()
    }

    Gdx.app.debug(TAG, userJson)

    println(userJson)
    val obj = JSONObject(userJson)
    println(obj.get("username"))
    characterArray = obj.getJSONArray("characters")
    setUpUI()
  }

  override fun setUpUI() {
    var player:GameObject? = null

    val mainTable = Table()
    stage.addActor(mainTable)
    mainTable.setFillParent(true)
    mainTable.debug = false

    val label = Label("Choose your character", fieldSkins)
    mainTable.add(label).padBottom(25f)
    mainTable.row()

    val charactersTable = Table()
    val scrollPane = ScrollPane(charactersTable, fieldSkins)
    mainTable.add(scrollPane).padBottom(50f)


    val seedGroup = Table()
    mainTable.add(seedGroup)

    val selectBox = SelectBox<String>(fieldSkins)
    seedGroup.add(selectBox)
    seedGroup.row().pad(15f)

    val submitButton = TextButton("Enter World", fieldSkins)
    seedGroup.add(submitButton)





    for ((i, character) in characterArray.withIndex()) {
      val playerData = character as JSONObject
      //val characterButton = TextButton(playerData.getString("name") + " " + playerData.get("characterClass"), fieldSkins)

      val characterTable = Table()
      characterTable.debug = false
      charactersTable.add(characterTable).pad(15f).align(-1)
      if((i+1) % 2 == 0){
        charactersTable.row()
      }
      val characterImage = Image(Assets.stateAnimations[playerData.getString("characterClass").toLowerCase()+playerData.getString("gender").toLowerCase()]!![State.IDLE]!!.getKeyFrame(0f))
      characterTable.add(characterImage).padRight(20f)

      val characterInfoGroup = Table()
      val characterName = Label(playerData.getString("name"), fieldSkins)
      characterInfoGroup.add(characterName).align(-1)
      characterInfoGroup.row()
      val characterClass = Label(playerData.getString("characterClass"), fieldSkins)
      characterInfoGroup.add(characterClass).align(-1)
      characterTable.add(characterInfoGroup)

      characterInfoGroup.addListener(object : ClickListener() {
        override fun touchDown(event: InputEvent?, x: Float, y: Float, point: Int, button: Int): Boolean {
          println("Picked " + playerData.getString("name"))
          val kClass = Class.forName("com.bosowski.oslark.gameObjects.prefabs.playerClasses.${playerData.getString("characterClass").toLowerCase().capitalize()}").kotlin
          player = kClass.constructors.first().call(playerData.getString("gender").toLowerCase()) as GameObject
          World.playerName = playerData.getString("name")

          //set up available seeds.
          val array = com.badlogic.gdx.utils.Array<String>()
          println(NetworkManager.instance.getScores(World.playerName!!))
          val scores = JSONObject(NetworkManager.instance.getScores(World.playerName!!)).getJSONArray("highscores")
          println(scores.toString())
          scores.forEach {
            val foundSeed = JSONObject(it.toString()).getLong("seed")
            if(!array.contains("#$foundSeed")){
              array.add("#$foundSeed")
            }
          }
          array.add("#"+array.size.toString() + " (new)")
          selectBox.items = array

          //set up the listener after some character is picked.
          submitButton.listeners.clear()
          submitButton.addListener(object : ClickListener() {
            override fun touchDown(event: InputEvent?, x: Float, y: Float, point: Int, button: Int): Boolean {
              if(player != null){
                game.screen = GameScreen(game, player!!, selectBox.selected.toString().replace("#","").replace(" (new)","").toLong())
                return true
              }
              return false
            }
          })
          return true
        }
      })
    }
  }

  override fun render(deltaTime: Float) {
    Gdx.gl.glClearColor(49f/255, 38f/255, 45f/255, 1f)
    Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT)

    stage.act(deltaTime)
    stage.draw()
  }

  override fun show() {
    Gdx.input.inputProcessor = stage
  }

  override fun hide() {}

  override fun pause() {}
}
